contracts/treasury/priceOracle/KyberDmmChainLinkPriceOracle.sol

```diff
@@ -0,0 +1,523 @@
+      if (_isAdd) {
+        _whitelistedTokens.add(_tokens[i]);
+      } else {
+        _whitelistedTokens.remove(_tokens[i]);
+      }
+      emit WhitelistedTokenUpdated(_tokens[i], _isAdd);
+    }
+  }
+
+  function _setDefaultPremiumData(
+    uint64 _removeLiquidityBps,
+    uint64 _liquidateLpBps,
+    uint64 _liquidateTokensBps
+  ) internal {
+    require(_removeLiquidityBps < BPS, 'invalid remove liquidity bps');
+    require(_liquidateLpBps < BPS, 'invalid liquidate lp bps');
+    require(_liquidateTokensBps < BPS, 'invalid liquidate tokens bps');
+    _defaultPremiumData.removeLiquidityBps = _removeLiquidityBps;
+    _defaultPremiumData.liquidateLpBps = _liquidateLpBps;
+    _defaultPremiumData.liquidateTokensBps = _liquidateTokensBps;
+    emit DefaultPremiumDataSet(_removeLiquidityBps, _liquidateLpBps, _liquidateTokensBps);
+  }
+
+  function _setGroupPremiumData(
+    address _liquidator,
+    uint64 _removeLiquidityBps,
+    uint64 _liquidateLpBps,
+    uint64 _liquidateTokensBps
+  ) internal {
+    require(_removeLiquidityBps < BPS, 'invalid remove liquidity bps');
+    require(_liquidateLpBps < BPS, 'invalid liquidate lp bps');
+    require(_liquidateTokensBps < BPS, 'invalid liquidate tokens bps');
+    _groupPremiumData[_liquidator].removeLiquidityBps = _removeLiquidityBps;
+    _groupPremiumData[_liquidator].liquidateLpBps = _liquidateLpBps;
+    _groupPremiumData[_liquidator].liquidateTokensBps = _liquidateTokensBps;
+    emit UpdateGroupPremiumData(_liquidator, _removeLiquidityBps, _liquidateLpBps, _liquidateTokensBps);
+  }
+
+  function _applyPremiumFor(uint256[] memory amounts, uint64 premiumBps)
+    internal pure
+    returns (uint256[] memory finalAmounts)
+  {
+    finalAmounts = amounts;
+    for(uint256 i = 0; i < finalAmounts.length; i++) {
+      finalAmounts[i] -= finalAmounts[i].mul(premiumBps) / BPS;
+    }
+  }
+
+  /**
+  *   @dev Get expected return amount from src token given dest token data
+  *   Save gas when liquidating multiple tokens or LP tokens
+  */
+  function _getExpectedReturnFromToken(
+    IERC20Ext tokenIn,
+    uint256 amountIn,
+    IERC20Ext dest,
+    uint256 destRateEth,
+    uint256 destRateUsd,
+    bool isFromLpToken
+  )
+    internal view
+    returns (uint256 totalReturn)
+  {
+    bool isDestEth = dest == ETH_TOKEN_ADDRESS || dest == IERC20Ext(weth);
+    uint256 rate;
+
+    if (!isFromLpToken) {
+      rate = isDestEth ? getRateOverEth(address(tokenIn)) :
+        _getRateWithDestTokenData(address(tokenIn), destRateEth, destRateUsd);
+      require(rate > 0, '0 aggregator rate');
+      return calculateReturnAmount(amountIn, getDecimals(tokenIn), getDecimals(dest), rate);
+    }
+
+    (IERC20Ext[2] memory tokens, uint256[2] memory amounts) = getExpectedTokensFromLp(
+      address(tokenIn), amountIn
+    );
+
+    uint256 destTokenDecimals = getDecimals(dest);
+
+    // calc equivalent (tokens[0], amounts[0]) -> tokenOuts[0]
+    if (tokens[0] == dest) {
```

Comment:
Checks like:
- if src == dest
- if dest == eth or weth
shold move in `_getRateWithDestTokenData` function